### 异常抛出
java异常分为受检、非受检异常。受检异常需要捕获并处理；非受检不受限制。
```java
// 获取stacktrace
public synchronized Throwable fillInStackTrace() {
if (stackTrace != null || backtrace != null) { 
	fillInStackTrace(0); // native 
	stackTrace = UNASSIGNED_STACK;
	}
	return this; 
}
// 获取stacktrace
private native Throwable fillInStackTrace(int dummy);
```
### 异常处理表
code属性
```
u2 exception_table_length; 
{
	u2 start_pc; 
	u2 end_pc;
	u2 handler_pc; 
	u2 catch_type;
} exception_table[exception_table_length];
```
- start_pc , end_pc 受检查之间的代码块索引范围
- catch_type：异常类常量池索引
- handler_pc：负责处理当前异常的catch块pc

1. 如果位于start_pc和end_pc之间的指令抛出异常x，且x是X(或者 X的子类)的实例。如果当前方法无法处理异常，则不断向上查看调用者是否可以处理。
2. 如果catch_type为0，表示catch-all
#### 异常抛出指令 athrow
需要一个操作数，从栈中弹出异常对象引用
```go
func (self *ATHROW) Execute(frame *rtda.Frame) {
	ex := frame.OperandStack().PopRef()
	if ex == nil {
		panic("java.lang.NullPointerException")
	}

	thread := frame.Thread()
	if !findAndGotoExceptionHandler(thread, ex) {
		handleUncaughtException(thread, ex)
	}
}

func findAndGotoExceptionHandler(thread *rtda.Thread, ex *heap.Object) bool {
	for {
		frame := thread.CurrentFrame()
		pc := frame.NextPC() - 1

		handlerPC := frame.Method().FindExceptionHandler(ex.Class(), pc)
		if handlerPC > 0 {
			stack := frame.OperandStack()
			stack.Clear() // 清空操作数栈
			stack.PushRef(ex)
			frame.SetNextPC(handlerPC)
			return true
		}

		thread.PopFrame()
		if thread.IsStackEmpty() {
			break
		}
	}
	return false
}

func handleUncaughtException(thread *rtda.Thread, ex *heap.Object) {
	thread.ClearStack() // 清空java虚拟机栈

	jMsg := ex.GetRefVar("detailMessage", "Ljava/lang/String;")
	goMsg := heap.GoString(jMsg)
	println(ex.Class().JavaName() + ": " + goMsg)

	stes := reflect.ValueOf(ex.Extra()) // ex.Extra()存放的是虚拟机栈帧信息，由native method fillInStackTrace塞入
	for i := 0; i < stes.Len(); i++ {
		ste := stes.Index(i).Interface().(interface {
			String() string
		})
		println("\tat " + ste.String())
	}

```